GNSSLogger/app/src/test/java/com/google/android/apps/location/gps/gnsslogger/PendingIntentCompatTest.java

```diff
@@ -0,0 +1,90 @@
+package com.google.android.apps.location.gps.gnsslogger;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.robolectric.Shadows.shadowOf;
+
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.os.Build;
+import android.os.Bundle;
+import androidx.test.core.app.ApplicationProvider;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+import org.robolectric.shadows.ShadowPendingIntent;
+
+/** Unit test for {@link PendingIntentCompat}. */
+@RunWith(RobolectricTestRunner.class)

```

Comment:
> 

Thanks for your advice. Robolectric is very lightweight compared to Expresso. My class is simple enough and does not require too complicated tests. I think it may easier to use Robolectric.

Reply:
> Could you use Expresso instead of Robolectric?
> 
> https://developer.android.com/training/testing/espresso/intents
> 
> I'd prefer to avoid external testing libraries unless absolutely necessary. I've opened PR #38 to add CI with multiple API levels that we can test on going forward. After that PR is merged, please merge master into this branch so the tests run on CI.
> 
> Also, my preference would be that all new classes be implemented in Kotlin instead of Java. See other comment for a way to do a quick conversion.

Thanks for your advice. Robolectric is very lightweight compared to Expresso. My class is simple enough and does not require too complicated tests. I think it may easier to use Robolectric.

Reply:
My issue with Robolectric is that it doesn't actually run on Android platform code, so you're not testing on an actual Android implementation. As a result I've preferred instrumented tests and have used Robolectric only as a method of last resort (e.g., when performance is critical, instrumented tests can't run on emulators/CI). GnssLogger will eventually have other instrumented tests so adding one more here for the PendingIntent doesn't add significant overhead.

It seems like the AndroidX library also prefers instrumented tests to Robolectric.

If you take a look, most of the unit tests for the AndroidX core library are here in `androidTest`:
https://github.com/androidx/androidx/tree/androidx-main/core/core/src/androidTest/java/androidx/core

...not in `test` using Robolectric on the JVM:
https://github.com/androidx/androidx/tree/androidx-main/core/core/src/test/java/androidx/core

If you look at the Robolectric tests, they are annotated with `@DoNotInstrument`, which means they exist there for a reason.

I also don't think you need full blown Expresso here - if you take a look at some of the Android X instrumented tests you'll get an idea of how you could mirror that approach with the PendingIntent.

Reply:
> My issue with Robolectric is that it doesn't actually run on Android platform code, so you're not testing on an actual Android implementation. As a result I've preferred instrumented tests and have used Robolectric only as a method of last resort (e.g., when performance is critical, instrumented tests can't run on emulators/CI). GnssLogger will eventually have other instrumented tests so adding one more here for the PendingIntent doesn't add significant overhead.
> 
> It seems like the AndroidX library also prefers instrumented tests to Robolectric.
> 
> If you take a look, most of the unit tests for the AndroidX core library are here in `androidTest`: https://github.com/androidx/androidx/tree/androidx-main/core/core/src/androidTest/java/androidx/core
> 
> ...not in `test` using Robolectric on the JVM: https://github.com/androidx/androidx/tree/androidx-main/core/core/src/test/java/androidx/core
> 
> If you look at the Robolectric tests, they are annotated with `@DoNotInstrument`, which means they exist there for a reason.
> 
> I also don't think you need full blown Expresso here - if you take a look at some of the Android X instrumented tests you'll get an idea of how you could mirror that approach with the PendingIntent.

I agree with that instrumented tests are needed in this project. However, I don't know how to apply it on `PendingIntentCompat`. Because we can't get the parameters information directly through creating `PendingIntent` object. https://github.com/androidx/androidx/blob/androidx-main/core/core/src/androidTest/java/androidx/core/app/DialogCompatTest.java. However, shadow of Robolectirc could help us get those information.